### ISSUE TITLE
Feature to 'Enqueue' Activities anywhere

### ISSUE THREAD

**Issue by thdotnet** [Azure/azure-functions-durable-extension#1107]
At this moment, only the function which receives 'IDurableOrchestrationContext' can start sub activities. It would be good if this could be injected into any Function, so we can start new activities inside a current activity. 

eg.

```
Function Orchestrator -> await ActivityA
                                     -> await ActivityB
                                                                - CallActivityAsync("B1")
                                                                - CallActivityAsync("B2")
```

The reason for that is because the sub activities B1 and B2 may or may not be executed, it depends on received params.

**Comment by cgillum** [comment]
One challenge with being able to call activity functions from a non-orchestrator function is that we won't have a simple way to return the response back to the calling function. Would it be acceptable if we only allowed one-way, fire-and-forget calls to activity functions from other non-orchestrator functions?

**Comment by thdotnet** [comment]
Yes and no. The reason I'm asking this is to avoid "Conditional" branches in the orchestrator when it's not needed (for business reasons). So, yes, it's will run on a separate task / function which will have it's own lifecycle. What do I mean by that, assuming we have a 30min execution for a function, if it's already spent 10 minutes, this fire and forget needs to have another 30min, not the remaining 20min from the parent one.

**Comment by thdotnet** [comment]
> we won't have a simple way to return the response back to the calling function

Is there any Stack or Queue under the hood or an Observable collection which the caller could be monitoring?

**Comment by cgillum** [comment]
> Is there any Stack or Queue under the hood or an Observable collection which the caller could be monitoring?

Maybe, but the challenge is correlating a particular message in that queue back to a particular client function execution.

Suppose for example that I have activity function A that calls activity function B.  Now suppose I invoke activity function A 10 times in parallel. If there is a single queue handling return messages, how do I guarantee that each return message from B is fed to the correct execution of A? There are many other problems that need to be considered too if an application is scaled out across multiple machines. I don't believe there is a simple way to accomplish this without using a TCP connection, and at that point you're probably better off either using a simple method call or an HTTP trigger.

Orchestrations can handle this properly because they are durable and addressable. A response message contains an instance ID and our dispatcher knows how to load that instance into memory and route it to the correct orchestrator function. All other function types, on the other hand, are stateless and transient and therefore .

This is an example of why I think we could do this for one-way calls but not for two-way calls. Does that help clarify?

> The reason I'm asking this is to avoid "Conditional" branches in the orchestrator when it's not needed (for business reasons). I mean, the flow in the orchestrator can be really messy.

I'm not sure I understand this. How does calling an activity function from another activity function remove unnecessary conditionals (and why are there unnecessary conditionals in the first place)? It might help if you could give me an example of some before/after code that illustrates what problem you're trying to solve. Depending on the problem, there may actually be other types of solutions we could explore.

**Comment by thdotnet** [comment]
Got it. So maybe the best is work with sub orchestrators paying attention to the restrictions (deterministic, non I/O, etc). Thanks for the explanations

> I'm not sure I understand this. How does calling an activity function from another activity function remove unnecessary conditionals (and why are there unnecessary conditionals in the first place)? It might help if you could give me an example of some before/after code that illustrates what problem you're trying to solve. Depending on the problem, there may actually be other types of solutions we could explore.

The if / branch will still exists, but not in the orchestrator. That was my point, I'm ok with that, but just would like to keep the orchestration logic simple.
